File: src-tauri/src/commands/model_verification.rs
use crate::{
    relay::model_verification::{
        target,
        types::{
            RunFailureKind, StartRunResponse, TargetKey, TargetScope, VerificationHistoryEntry,
            VerificationReport,
        },
        verdict::report_precedes,
    },
    AppState,
};
use serde::Serialize;

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct VerificationScopeSummary {
    pub provider_id: String,
    pub app_type: String,
    pub badge_verdict: Option<crate::relay::model_verification::types::Verdict>,
    pub representative_report: VerificationReport,
}

#[tauri::command]
pub async fn list_verification_models(
    state: tauri::State<'_, AppState>,
    provider_id: String,
    app_type: String,
) -> Result<Vec<crate::relay::model_verification::types::VerificationModelOption>, String> {
    target::list_models(&state.db, &provider_id, &app_type)
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn get_model_verification_diagnostics(
    state: tauri::State<'_, AppState>,
    provider_id: String,
    app_type: String,
    model: String,
) -> Result<Vec<crate::relay::model_verification::types::ProbeDiagnostic>, String> {
    crate::relay::model_verification::store::active_diagnostics(
        &state.db,
        &provider_id,
        &app_type,
        &model,
    )
    .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn start_model_verification(
    state: tauri::State<'_, AppState>,
    provider_id: String,
    app_type: String,
    model: String,
) -> Result<StartRunResponse, RunFailureKind> {
    start_model_verification_impl(&state, provider_id, app_type, model).await
}

async fn start_model_verification_impl(
    state: &AppState,
    provider_id: String,
    app_type: String,
    model: String,
) -> Result<StartRunResponse, RunFailureKind> {
    state
        .model_verification
        .start(TargetKey::new(provider_id, app_type, model))
        .await
}

#[tauri::command]
pub fn cancel_model_verification(
    state: tauri::State<'_, AppState>,
    run_id: String,
) -> Result<(), RunFailureKind> {
    cancel_model_verification_impl(&state, run_id)
}

fn cancel_model_verification_impl(state: &AppState, run_id: String) -> Result<(), RunFailureKind> {
    state.model_verification.cancel(&run_id)
}

#[tauri::command]
pub fn get_model_verification_summaries(
    state: tauri::State<'_, AppState>,
    provider_ids: Vec<String>,
    app_type: String,
) -> Result<Vec<VerificationScopeSummary>, RunFailureKind> {
    get_model_verification_summaries_impl(&state, provider_ids, app_type)
}

fn get_model_verification_summaries_impl(
    state: &AppState,
    provider_ids: Vec<String>,
    app_type: String,
) -> Result<Vec<VerificationScopeSummary>, RunFailureKind> {
    let reports = state
        .model_verification
        .list_results_for_app(&app_type, &provider_ids)?;
    let mut summaries = Vec::<VerificationScopeSummary>::new();
    for report in reports {
        if let Some(summary) = summaries
            .iter_mut()
            .find(|summary| summary.provider_id == report.target.provider_id)
        {
            if report_precedes(&report, &summary.representative_report) {
                summary.badge_verdict = badge_verdict(report.verdict);
                summary.representative_report = report;
            }
        } else {
            summaries.push(VerificationScopeSummary {
                provider_id: report.target.provider_id.clone(),
                app_type: report.target.app_type.clone(),
                badge_verdict: badge_verdict(report.verdict),
                representative_report: report,
            });
        }
    }
    Ok(summaries)
}

const fn badge_verdict(
    verdict: crate::relay::model_verification::types::Verdict,
) -> Option<crate::relay::model_verification::types::Verdict> {
    use crate::relay::model_verification::types::Verdict;
    match verdict {
        Verdict::Trusted | Verdict::Suspicious | Verdict::Anomaly => Some(verdict),
        Verdict::Inconclusive => None,
    }
}

#[tauri::command]
pub fn get_model_verification_history(
    state: tauri::State<'_, AppState>,
    provider_id: String,
    app_type: String,
) -> Result<Vec<VerificationHistoryEntry>, RunFailureKind> {
    state
        .model_verification
        .list_history(&TargetScope::new(provider_id, app_type))
}

#[cfg(test)]
mod tests {
    use std::sync::{Arc, Mutex};

    use crate::{
        database::Database,
        relay::model_verification::{
            coordinator::{
                ActiveVerifier, ModelVerificationCoordinator, PreparedVerification, ProbeProgress,
            },
            store::upsert_active,
            types::{
                EvidenceCode, EvidenceFact, EvidenceLevel, EvidenceOutcome, RunFailureKind,
                TargetKey, Verdict, VerificationReport, RULES_VERSION,
            },
        },
        AppState,
    };

    use super::{
        cancel_model_verification_impl, get_model_verification_summaries_impl,
        start_model_verification_impl, VerificationScopeSummary,
    };

    struct RejectingVerifier {
        target: Mutex<Option<TargetKey>>,
    }

    impl ActiveVerifier for RejectingVerifier {
        fn prepare(
            &self,
            target: TargetKey,
            _progress: ProbeProgress,
        ) -> Result<PreparedVerification, RunFailureKind> {
            *self.target.lock().unwrap() = Some(target);
            Err(RunFailureKind::Authentication)
        }
    }

    fn state_with_verifier(verifier: Arc<dyn ActiveVerifier>) -> AppState {
        let db = Arc::new(Database::memory().unwrap());
        let mut state = AppState::new(db.clone());
        state.model_verification =
            Arc::new(ModelVerificationCoordinator::with_verifier(db, verifier));
        state
    }

    #[tokio::test]
    async fn start_boundary_passes_only_identifiers_and_returns_a_finite_failure() {
        let verifier = Arc::new(RejectingVerifier {
            target: Mutex::new(None),
        });
        let state = state_with_verifier(verifier.clone());

        let failure = start_model_verification_impl(
            &state,
            "provider-a".into(),
            "codex".into(),
            "gpt-5.6-sol".into(),
        )
        .await
        .unwrap_err();

        assert_eq!(failure, RunFailureKind::Authentication);
        assert_eq!(
            verifier.target.lock().unwrap().as_ref(),
            Some(&TargetKey::new("provider-a", "codex", "gpt-5.6-sol"))
        );
        let serialized = serde_json::to_string(&failure).unwrap();
        for sentinel in ["URL", "KEY", "PROMPT", "OUTPUT", "THINKING", "SIGNATURE"] {
            assert!(!serialized.contains(sentinel));
        }
    }

    #[tokio::test]
    async fn cancel_unknown_run_and_empty_result_query_are_idempotent() {
        let verifier = Arc::new(RejectingVerifier {
            target: Mutex::new(None),
        });
        let state = state_with_verifier(verifier);

        cancel_model_verification_impl(&state, "unknown-run".into()).unwrap();
        assert!(
            get_model_verification_summaries_impl(&state, Vec::new(), "codex".into())
                .unwrap()
                .is_empty()
        );
    }

    #[test]
    fn result_boundary_preserves_persisted_finite_evidence() {
        let verifier = Arc::new(RejectingVerifier {
            target: Mutex::new(None),
        });
        let state = state_with_verifier(verifier);
        state
            .db
            .conn
            .lock()
            .unwrap()
            .execute(
                "INSERT INTO providers (id, app_type, name, settings_config) VALUES (?1, ?2, ?3, '{}')",
                rusqlite::params!["provider-a", "codex", "provider-a"],
            )
            .unwrap();
        let fact = EvidenceFact {
            code: EvidenceCode::ModelMatch,
            outcome: EvidenceOutcome::Passed,
        };
        upsert_active(
            &state.db,
            &VerificationReport {
                target: TargetKey::new("provider-a", "codex", "gpt-a"),
                verdict: Verdict::Trusted,
                evidence_level: EvidenceLevel::ProtocolBehavior,
                facts: vec![fact.clone()],
                rules_version: RULES_VERSION,
                checked_at: 1_700_000_000,
            },
        )
        .unwrap();

        let results = get_model_verification_summaries_impl(
            &state,
            vec!["provider-a".into()],
            "codex".into(),
        )
        .unwrap();

        assert_eq!(
            results,
            vec![VerificationScopeSummary {
                provider_id: "provider-a".into(),
                app_type: "codex".into(),
                badge_verdict: Some(Verdict::Trusted),
                representative_report: VerificationReport {
                    target: TargetKey::new("provider-a", "codex", "gpt-a"),
                    verdict: Verdict::Trusted,
                    evidence_level: EvidenceLevel::ProtocolBehavior,
                    facts: vec![fact],
                    rules_version: RULES_VERSION,
                    checked_at: 1_700_000_000,
                },
            }]
        );
    }

    #[test]
    fn result_boundary_aggregates_scope_severity_and_uses_the_newest_tie() {
        let verifier = Arc::new(RejectingVerifier {
            target: Mutex::new(None),
        });
        let state = state_with_verifier(verifier);
        state
            .db
            .conn
            .lock()
            .unwrap()
            .execute(
                "INSERT INTO providers (id, app_type, name, settings_config) VALUES (?1, ?2, ?3, '{}')",
                rusqlite::params!["provider-a", "codex", "provider-a"],
            )
            .unwrap();
        for (model, verdict, checked_at) in [
            ("trusted", Verdict::Trusted, 30),
            ("old-anomaly", Verdict::Anomaly, 10),
            ("new-anomaly", Verdict::Anomaly, 20),
        ] {
            upsert_active(
                &state.db,
                &VerificationReport {
                    target: TargetKey::new("provider-a", "codex", model),
                    verdict,
                    evidence_level: EvidenceLevel::ProtocolBehavior,
                    facts: Vec::new(),
                    rules_version: RULES_VERSION,
                    checked_at,
                },
            )
            .unwrap();
        }

        let summaries = get_model_verification_summaries_impl(
            &state,
            vec!["provider-a".into()],
            "codex".into(),
        )
        .unwrap();

        assert_eq!(summaries.len(), 1);
        assert_eq!(summaries[0].badge_verdict, Some(Verdict::Anomaly));
        assert_eq!(
            summaries[0].representative_report.target.model,
            "new-anomaly"
        );
    }
}

File: src/lib/api/modelVerification.ts
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import { invoke } from "@tauri-apps/api/core";

import {
  MODEL_VERIFICATION_CHANGED,
  MODEL_VERIFICATION_PROGRESS,
} from "./events";

export type VerificationVerdict =
  "trusted" | "suspicious" | "anomaly" | "inconclusive";

export type EvidenceLevel =
  "cryptographic" | "protocolBehavior" | "insufficient";

export type EvidenceCode =
  | "basicEnvelope"
  | "modelMatch"
  | "streamLifecycle"
  | "usageConsistency"
  | "toolCallShape"
  | "structuredOutput"
  | "thinkingSignature"
  | "signatureContinuation"
  | "foreignProtocol"
  | "foreignSelfIdentification";

export type EvidenceOutcome = "passed" | "failed" | "skipped";

export type RunFailureKind =
  | "authentication"
  | "rateLimited"
  | "insufficientBalance"
  | "network"
  | "upstream"
  | "timeout"
  | "modelUnavailable"
  | "cancelled"
  | "invalidResponse";

export type RunState =
  "queued" | "running" | "completed" | "cancelled" | "failed";

export interface VerificationTarget {
  providerId: string;
  appType: string;
  model: string;
}

export interface VerificationScope {
  providerId: string;
  appType: string;
}

export interface EvidenceFact {
  code: EvidenceCode;
  outcome: EvidenceOutcome;
}

export interface VerificationReport {
  target: VerificationTarget;
  verdict: VerificationVerdict;
  evidenceLevel: EvidenceLevel;
  facts: EvidenceFact[];
  rulesVersion: number;
  checkedAt: number;
}

/** 失败腿的原始请求/响应（诊断边车，只在证据 Failed 时有）。 */
export interface ProbeDiagnostic {
  /** 产出该事实的探针腿：core / identity / tool / structured / stream。 */
  probe: string;
  code: EvidenceCode;
  request: string;
  response: string;
}

export interface VerificationScopeSummary {
  providerId: string;
  appType: string;
  /** 后端决定是否在档位行展示；不确定结论为 null。 */
  badgeVerdict: VerificationVerdict | null;
  /** 后端按严重度、同级最新时间选出的弹窗默认报告。 */
  representativeReport: VerificationReport;
}

export type VerificationSource = "active";

export interface VerificationHistoryEntry {
  source: VerificationSource;
  report: VerificationReport;
}

export interface StartRunResponse {
  runId: string;
  state: RunState;
}

export interface VerificationProgressEvent extends VerificationTarget {
  runId: string;
  state: RunState;
  completedChecks: number;
  totalChecks: number;
  failure: RunFailureKind | null;
}

/** 协议预筛结论（后端 `ModelFitness` 的镜像；判据唯一源是站点 ai-transit 快照）。 */
export type ModelFitness = "supported" | "unsupported_protocol" | "unknown";

/** 验证弹窗的一个模型选项。`unsupported` 项禁选并带原因标签。 */
export interface VerificationModelOption {
  name: string;
  fitness: ModelFitness;
}

export const modelVerificationApi = {
  listModels: (
    providerId: string,
    appType: string,
  ): Promise<VerificationModelOption[]> =>
    invoke("list_verification_models", { providerId, appType }),

  diagnostics: (
    providerId: string,
    appType: string,
    model: string,
  ): Promise<ProbeDiagnostic[]> =>
    invoke("get_model_verification_diagnostics", {
      providerId,
      appType,
      model,
    }),

  start: (target: VerificationTarget): Promise<StartRunResponse> =>
    invoke("start_model_verification", {
      providerId: target.providerId,
      appType: target.appType,
      model: target.model,
    }),

  cancel: (runId: string): Promise<void> =>
    invoke("cancel_model_verification", { runId }),

  listSummaries: (
    providerIds: string[],
    appType: string,
  ): Promise<VerificationScopeSummary[]> =>
    invoke("get_model_verification_summaries", { providerIds, appType }),

  listHistory: (
    providerId: string,
    appType: string,
  ): Promise<VerificationHistoryEntry[]> =>
    invoke("get_model_verification_history", { providerId, appType }),

  onProgress: (
    handler: (event: VerificationProgressEvent) => void,
  ): Promise<UnlistenFn> =>
    listen<VerificationProgressEvent>(MODEL_VERIFICATION_PROGRESS, (event) =>
      handler(event.payload),
    ),

  onChanged: (
    handler: (scope: VerificationScope) => void,
  ): Promise<UnlistenFn> =>
    listen<VerificationScope>(MODEL_VERIFICATION_CHANGED, (event) =>
      handler(event.payload),
    ),
};
